inc/functions.php

```diff
@@ -566,47 +624,73 @@ function my_date($format, $stamp=0, $offset="", $ty=1, $adodb=false)
  */
 function my_mail($to, $subject, $message, $from="", $charset="", $headers="", $keep_alive=false, $format="text", $message_text="", $return_email="")
 {
-	global $mybb;
+	global $mybb, $plugins;
 	static $mail;
 
-	// Does our object not exist? Create it
-	if(!is_object($mail))
-	{
-		require_once MYBB_ROOT."inc/class_mailhandler.php";
+	// Get our mail handler.
+	$mail = &get_mybb_mailhandler();
 
-		if($mybb->settings['mail_handler'] == 'smtp')
-		{
-			require_once MYBB_ROOT."inc/mailhandlers/smtp.php";
-			$mail = new SmtpMail();
-		}
-		else
-		{
-			require_once MYBB_ROOT."inc/mailhandlers/php.php";
-			$mail = new PhpMail();
-		}
-	}
-
-	// Using SMTP based mail
-	if($mybb->settings['mail_handler'] == 'smtp')
+	// If MyBB's built-in SMTP mail handler is used, set the keep alive bit accordingly.
+	if(isset($mybb->settings['mail_handler']) && $mybb->settings['mail_handler'] == 'smtp' && $keep_alive == true && !empty($mail->keep_alive))
 	{
-		if($keep_alive == true)
+		require_once MYBB_ROOT . "inc/class_mailhandler.php";
+		require_once MYBB_ROOT . "inc/mailhandlers/smtp.php";
+		if($mail instanceof MailHandler && $mail instanceof SmtpMail)
```

Comment:
Isn't `instanceof MailHandler` already checked above?

Reply:
No it isn't.